server/app/controllers/applicant/ApplicantProgramBlocksController.java

```diff
@@ -657,46 +659,102 @@ public CompletionStage<Result> updateWithApplicantId(
+                v -> applicantService.getReadOnlyApplicantProgramService(applicantId, programId),
+                classLoaderExecutionContext.current())
+            .toCompletableFuture();
+
+    return CompletableFuture.allOf(
+            formDataCompletableFuture, applicantProgramServiceCompletableFuture)
         .thenComposeAsync(
-            formData ->
-                applicantService.stageAndUpdateIfValid(
+            (v) -> {
+              ImmutableMap<String, String> formData = formDataCompletableFuture.join();
+              ReadOnlyApplicantProgramService readOnlyApplicantProgramService =
+                  applicantProgramServiceCompletableFuture.join();
+              CiviFormProfile profile = profileUtils.currentUserProfileOrThrow(request);
+              Optional<Block> maybeBlockBeforeUpdate =
+                  readOnlyApplicantProgramService.getActiveBlock(blockId);
+              ApplicantRequestedAction applicantRequestedAction =
+                  applicantRequestedActionWrapper.getAction();
+
+              // Allow applicants to immediately navigate away from a block if they haven't even
+              // started answering it. We can immediately navigate away from a block if:
+              if (
+              // There are no answers currently filled out...
+              formData.values().stream().allMatch(value -> value.equals(""))
+                  // ... and the applicant wants to navigate to previous or review...
+                  // [Explanation: We can't let applicants navigate to the next block without
+                  // answers because the next block may have a visibility conditions that
+                  // depends on the answers to this block.]
+                  && (applicantRequestedAction == REVIEW_PAGE
+                      || applicantRequestedAction == PREVIOUS_BLOCK)
+                  && maybeBlockBeforeUpdate.isPresent()
+                  // ... and the applicant didn't previously complete this block...
+                  && !maybeBlockBeforeUpdate.get().isCompletedInProgramWithoutErrors()
+                  // ...and there's at least one required question, meaning that all blank answers
+                  // is *not* a valid response.
+                  // [Explanation: We don't allow applicants to submit an application until they've
+                  // at least seen all the questions. We mark as question as "seen" by checking
+                  // that the metadata is filled in (see
+                  // {@link ApplicantQuestion#isAnsweredOrSkippedOptionalInProgram}).
+                  //
+                  // If a block has all optional questions, having all empty answers is a valid
+                  // response and we should mark that block as seen by having
+                  // {@link #stageAndUpdateIfValid} fill in that metadata. So, we can't immediately
+                  // navigate away.]
+                  && !maybeBlockBeforeUpdate.get().hasOnlyOptionalQuestions()) {
```

Comment:
(blocking) This `if` statement is getting pretty long and complicated. The documentation helps immensely, but please move it into a function of it's own called `canNavigateAwayFromPageImmediately` or something along those lines.